presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/StateSerializer.java

```diff
@@ -0,0 +1,282 @@
+            return new WeightedReservoirSampleStateStrategy(size);
+        }
+
+        throw new InvalidParameterException(format("unknown method %s or method/argument mismatch", method));
+    }
+
+    public static StateStrategy create(
+            long size,
+            double sample,
+            double weight,
+            String method,
+            double min,
+            double max)
+    {
+        if (method.equalsIgnoreCase("fixed_histogram_mle")) {
+            return new FixedHistogramMLEStateStrategy(size, min, max);
+        }
+
+        if (method.equalsIgnoreCase("fixed_histogram_jacknife")) {
+            return new FixedHistogramJacknifeStateStrategy(size, min, max);
+        }
+
+        throw new InvalidParameterException(format("unknown method %s or method/argument mismatch", method));
+    }
+
+    public static void validate(StateStrategy strategy, String method)
+    {
+        if (method.equalsIgnoreCase("reservoir_vasicek")) {
+            if (strategy instanceof UnweightedReservoirSampleStateStrategy) {
+                return;
+            }
+
+            if (strategy instanceof WeightedReservoirSampleStateStrategy) {
+                return;
+            }
+
+            throw new PrestoException(
+                    INVALID_FUNCTION_ARGUMENT,
+                    "Inconsistent method");
+        }
+
+        if (method.equalsIgnoreCase("weighted_reservoir")) {
+            if (!(strategy instanceof WeightedReservoirSampleStateStrategy)) {
+                throw new PrestoException(
+                        INVALID_FUNCTION_ARGUMENT,
+                        "Inconsistent method");
+            }
+
+            return;
+        }
+
+        if (method.equalsIgnoreCase("fixed_histogram_mle")) {
+            if (!(strategy instanceof FixedHistogramMLEStateStrategy)) {
+                throw new PrestoException(
+                        INVALID_FUNCTION_ARGUMENT,
+                        "Inconsistent method");
+            }
+
+            return;
+        }
+
+        if (method.equalsIgnoreCase("fixed_histogram_jacknife")) {
+            if (!(strategy instanceof FixedHistogramJacknifeStateStrategy)) {
+                throw new PrestoException(
+                        INVALID_FUNCTION_ARGUMENT,
+                        "Inconsistent method");
+            }
+
+            return;
+        }
+        throw new InvalidParameterException("unknown method");
+    }
+
+    public static void validate(StateStrategy strategy)
+    {
+        if (strategy instanceof UnweightedReservoirSampleStateStrategy) {
+            return;
+        }
+
+        if (strategy instanceof WeightedReservoirSampleStateStrategy) {
+            return;
+        }
+
+        throw new InvalidParameterException("Unsupported");
+    }
+
+    public static void combine(StateStrategy target, StateStrategy source)
+    {
+        if (target instanceof UnweightedReservoirSampleStateStrategy) {
```

Comment:
perhaps, simplify this method by replacing different if blocks with just one:

```
    private static void combine(StateStrategy target, StateStrategy source)
    {
        if (target.getClass() != source.getClass()) {
            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "<TBD: Write a good error message>");
        }
        target.mergeWith(source);
    }
```

Or, better yet add checks to mergeWith methods and remove the check here; then inline this method:

```
    @Override
    public void mergeWith(StateStrategy other)
    {
        checkArgument(other instanceof FixedHistogramJacknifeStateStrategy, "Unexpected strategy: " + other.getClass().getSimpleName());
        histogram.mergeWith(((FixedHistogramJacknifeStateStrategy) other).histogram);
    }

        if (strategy != null && otherStrategy != null) {
            strategy.mergeWith(otherStrategy);
        }
```